backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/FunctionReferenceLowering.kt

```diff
@@ -94,6 +94,34 @@ internal class CallableReferenceLowering(val context: Context): FileLoweringPass
                 return result
             }
 
+            // Handle SAM conversions which wrap a function reference:
+            //     class sam$n(private val receiver: R) : Interface { override fun method(...) = receiver.target(...) }
+            //
+            // This is actually very common, as `Interface { something }` is a local function + a SAM-conversion
+            // of a reference to it into an implementation.
+            //
+            // This is needed not only for optimization purposes (producing one less class)
+            // but also for implementing correct equals/hashCode methods.
+            override fun visitTypeOperator(expression: IrTypeOperatorCall): IrExpression {
+                if (expression.operator == IrTypeOperator.SAM_CONVERSION) {
+                    val invokable = expression.argument
+                    val reference = if (invokable is IrFunctionReference) {
+                        invokable
+                    } else if (invokable is IrBlock && (invokable.origin.isLambda)
+                            && invokable.statements.last() is IrFunctionReference) {
+                        val statements = invokable.statements
+                        for (i in 0 until statements.size - 1)
```

Comment:
Btw, the code below returns `transformFunctionReference(reference, expression.typeOperand)`. Does it drop all statements but last in this case then?

Reply:
Huh, indeed, this means that there is a dependency on LocalFunctionsLowering here. Because actually there are exactly two statements here: IrComposite (that was left of IrFunction) and IrFunctionReference.

Reply:
Please either don't rely on this here or check this explicitly (with proper comment).